tt_eager/tt_dnn/op_library/rotary_embedding/multi_core/rotary_embedding_llama_op_multi_core.cpp

```diff
@@ -0,0 +1,305 @@
+
+    const auto &cores = grid_to_cores(num_cores, num_cores_x, num_cores_y, row_major);
+
+    uint32_t num_cores_per_sin_cos_row = max((uint32_t) 1, (uint32_t)(num_cores / Ht)); // since sin/cos matrices have Ht rows
+    uint32_t core_idx = 0;
+    /*
+        Outer loop: Ht / num_sin_cos_rows_per_core iterations
+        Inner loop: # total cores / Ht iterations
+
+        Overall: # total cores iterations
+    */
+
+    for (uint32_t sin_cos_row = 0; sin_cos_row < Ht; sin_cos_row+=num_sin_cos_rows_per_core) {
+        uint32_t anchor_row = sin_cos_row;
+        for (uint32_t i = 0; i < num_cores_per_sin_cos_row; i++) {
+            const CoreCoord &core = cores.at(core_idx);
+
+            // Reader runtime args
+            uint32_t start_row = anchor_row + (i * num_rows_per_core * Ht); // anchor_row + stride
+            std::vector<uint32_t> reader_rt_args;
+            reader_rt_args = {
+                    src_buffer->address(),
+                    cos_buffer->address(),
+                    sin_buffer->address(),
+                    trans_mat_buffer->address(),
+                    start_row, // start_row_id
+                    sin_cos_row * Wt, // This range of this idx must be [0, HtWt - 1], where HtWt is the size of the sin/cos matrices in # of tiles
+            };
+            tt_metal::SetRuntimeArgs(program, unary_reader_kernel_id, core, reader_rt_args);
+
+
+            // Writer runtime args
+            tt_metal::SetRuntimeArgs(
+                program,
+                unary_writer_kernel_id,
+                core,
+                {dst_buffer->address(), start_row});
+
+            // Go to next core
+            core_idx++;
+        }
+    }
+
+    auto override_runtime_arguments_callback = [unary_reader_kernel_id,
+                                                unary_writer_kernel_id,
+                                                cores,
+                                                num_rows_per_core,
+                                                Wt
+                                                ](
+                                                   const void *operation,
+                                                   const Program &program,
+                                                   const std::vector<Tensor> &input_tensors,
+                                                   const std::vector<std::optional<const Tensor>> &,
+                                                   const std::vector<Tensor> &output_tensors) {
+
+        auto src_buffer = input_tensors.at(0).buffer();
+        auto cos_buffer = input_tensors.at(1).buffer();
+        auto sin_buffer = input_tensors.at(2).buffer();
+        auto trans_mat_buffer = input_tensors.at(3).buffer();
+
+        auto dst_buffer = output_tensors.at(0).buffer();
+
+        for (uint32_t i = 0, num_tiles_written = 0; i < cores.size(); ++i) {
+            const CoreCoord &core = cores.at(i);
+            {
```

Comment:
This override runtime args call is going to be slow because `GetRuntimeArgs` is called per core. The performant method of doing this is to call `GetRuntimeArgs` on the kernel_id, and then index into it per core. See this for an example: https://github.com/tenstorrent/tt-metal/blob/main/tt_eager/tt_dnn/op_library/eltwise_binary/multi_core/eltwise_binary_op_multi_core.cpp#L127